File: lib/models/banking.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'banking.freezed.dart';
part 'banking.g.dart';

@freezed
class Wallet with _$Wallet {
  const factory Wallet({
    required String playerId,
    required int pocketBalance,
    required int bankBalance,
    required DateTime updatedAt,
  }) = _Wallet;

  factory Wallet.fromJson(Map<String, dynamic> json) => _$WalletFromJson(json);
}

@freezed
class Transaction with _$Transaction {
  const factory Transaction({
    required int id,
    required DateTime createdAt,
    String? actorId,
    String? senderId,
    String? receiverId,
    required String source,
    required String destination,
    required int amount,
    required String kind,
    String? memo,
    String? idempotencyKey,
  }) = _Transaction;

  factory Transaction.fromJson(Map<String, dynamic> json) => _$TransactionFromJson(json);
}

@freezed
class InterestOffer with _$InterestOffer {
  const factory InterestOffer({
    required int id,
    required String playerId,
    required String forDate,
    required int bankBalanceSnapshot,
    required int rateBps,
    required int amount,
    required DateTime claimDeadline,
    DateTime? claimedAt,
    required DateTime createdAt,
  }) = _InterestOffer;

  factory InterestOffer.fromJson(Map<String, dynamic> json) => _$InterestOfferFromJson(json);
}

@freezed
class Player with _$Player {
  const factory Player({
    required String id,
    required String username,
    required DateTime createdAt,
  }) = _Player;

  factory Player.fromJson(Map<String, dynamic> json) => _$PlayerFromJson(json);
}

@freezed
class LedgerEntry with _$LedgerEntry {
  const factory LedgerEntry({
    required int id,
    required DateTime createdAt,
    required String kind,
    required int amount,
    required int delta,
    required String source,
    required String destination,
    String? counterpartyUsername,
    String? memo,
  }) = _LedgerEntry;

  factory LedgerEntry.fromJson(Map<String, dynamic> json) => _$LedgerEntryFromJson(json);
}

@freezed
class ApiResponse with _$ApiResponse {
  const factory ApiResponse({
    required bool success,
    Map<String, dynamic>? data,
    String? error,
    String? message,
  }) = _ApiResponse;

  factory ApiResponse.fromJson(Map<String, dynamic> json) => _$ApiResponseFromJson(json);
}

@freezed
class BalanceResponse with _$BalanceResponse {
  const factory BalanceResponse({
    required int pocketBalance,
    required int bankBalance,
    required int txId,
  }) = _BalanceResponse;

  factory BalanceResponse.fromJson(Map<String, dynamic> json) => _$BalanceResponseFromJson(json);
}

@freezed
class TransferResponse with _$TransferResponse {
  const factory TransferResponse({
    required int pocketBalance,
    required int bankBalance,
    required TransferReceipt receipt,
  }) = _TransferResponse;

  factory TransferResponse.fromJson(Map<String, dynamic> json) => _$TransferResponseFromJson(json);
}

@freezed
class TransferReceipt with _$TransferReceipt {
  const factory TransferReceipt({
    required String counterparty,
    required int amount,
    required String source,
    required String destination,
  }) = _TransferReceipt;

  factory TransferReceipt.fromJson(Map<String, dynamic> json) => _$TransferReceiptFromJson(json);
}

@freezed
class InterestClaimResponse with _$InterestClaimResponse {
  const factory InterestClaimResponse({
    required int bankBalance,
    required int claimedAmount,
  }) = _InterestClaimResponse;

  factory InterestClaimResponse.fromJson(Map<String, dynamic> json) => _$InterestClaimResponseFromJson(json);
}

File: backend/src/types/banking.ts
export interface Player {
  id: string;
  username: string;
  created_at: Date;
}

export interface Wallet {
  player_id: string;
  pocket_balance: number;
  bank_balance: number;
  updated_at: Date;
}

export interface Transaction {
  id: number;
  created_at: Date;
  actor_id?: string;
  sender_id?: string;
  receiver_id?: string;
  source: 'POCKET' | 'BANK';
  destination: 'POCKET' | 'BANK';
  amount: number;
  kind: 'DEPOSIT' | 'WITHDRAW' | 'TRANSFER_SEND' | 'TRANSFER_RECEIVE' | 'INTEREST_CLAIM' | 'ADMIN_ADJUST';
  memo?: string;
  idempotency_key?: string;
}

export interface InterestOffer {
  id: number;
  player_id: string;
  for_date: string;
  bank_balance_snapshot: number;
  rate_bps: number;
  amount: number;
  claim_deadline: Date;
  claimed_at?: Date;
  created_at: Date;
}

export interface DepositRequest {
  amount: number;
  memo?: string;
  idempotencyKey: string;
}

export interface WithdrawRequest {
  amount: number;
  memo?: string;
  idempotencyKey: string;
}

export interface TransferRequest {
  source: 'POCKET' | 'BANK';
  toUsername: string;
  amount: number;
  memo?: string;
  idempotencyKey: string;
}

export interface InterestClaimRequest {
  offerId: number;
  idempotencyKey: string;
}

export interface LedgerEntry {
  id: number;
  created_at: Date;
  kind: string;
  amount: number;
  delta: number; // + or - for display
  source: string;
  destination: string;
  counterparty_username?: string;
  memo?: string;
}

export interface AdminLedgerFilters {
  sender?: string;
  receiver?: string;
  from?: string;
  to?: string;
  kind?: string;
  min?: number;
  max?: number;
  limit?: number;
  cursor?: string;
}

export interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

export interface BalanceResponse {
  pocketBalance: number;
  bankBalance: number;
  txId: number;
}

export interface TransferResponse {
  pocketBalance: number;
  bankBalance: number;
  receipt: {
    counterparty: string;
    amount: number;
    source: string;
    destination: string;
  };
}

export interface InterestClaimResponse {
  bankBalance: number;
  claimedAmount: number;
}
